kuma/wiki/content.py

```diff
@@ -50,6 +49,117 @@
 DOC_SPECIAL_PATHS = ('new', 'tag', 'feeds', 'templates', 'needs-review')
 
 
+class Extractor(object):
+
+    def __init__(self, document):
+        self.document = document
+
+    def section(self, content, section_id, ignore_heading=False):
+        parsed_content = parse(content)
+        extracted = parsed_content.extractSection(section_id,
+                                                  ignore_heading=ignore_heading)
+        return extracted.serialize()
+
+    @newrelic.agent.function_trace()
+    def macro_names(self):
+        """
+        Extract a unique set of KumaScript macro names used in the content
+        """
+        names = set()
+        try:
+            txt = []
+            for token in parse(self.document.html).stream:
+                if token['type'] in ('Characters', 'SpaceCharacters'):
+                    txt.append(token['data'])
+            txt = ''.join(txt)
+            names.update(MACRO_RE.findall(txt))
+        except:
+            pass
+        return list(names)
+
+    @newrelic.agent.function_trace()
+    def css_classnames(self):
+        """
+        Extract the unique set of class names used in the content
+        """
+        classnames = set()
+        for element in pq(self.document.rendered_html).find('*'):
+            css_classes = element.attrib.get('class')
+            if css_classes:
+                classnames.update(css_classes.split(' '))
+        return list(classnames)
+
+    @newrelic.agent.function_trace()
+    def html_attributes(self):
+        """
+        Extract the unique set of HTML attributes used in the content
+        """
+        try:
+            attribs = []
+            for token in parse(self.document.rendered_html).stream:
+                if token['type'] == 'StartTag':
+                    for (namespace, name), value in token['data'].items():
+                        attribs.append((name, value))
+            return ['%s="%s"' % (k, v) for k, v in attribs]
+        except:
+            return []
+
+    @newrelic.agent.function_trace()
+    def code_sample(self, name):
+        """
+        Extract a dict containing the html, css, and js listings for a given
+        code sample identified by a name.
+
+        This should be pretty agnostic to markup patterns, since it just
+        requires a parent container with an DID and 3 child elements somewhere
+        within with class names "html", "css", and "js" - and our syntax
+        highlighting already does that with <pre>'s
+
+        Given the name of a code sample, attempt to extract it from rendered
+        HTML with a fallback to non-rendered in case of errors.
+        """
+        parts = ('html', 'css', 'js')
+        data = dict((x, None) for x in parts)
+
+        try:
+            src, errors = self.document.get_rendered()
```

Comment:
I'm wrong about ``{{EmbedLiveSample}}``.  It adds an iframe whose source is ``$samples/(sample_name)``, which renders the HTML document with the CSS, JS, and body HTML.  I still think there's latent bugs in here (need to test what happens when rendered_html is an empty string), but it should match current behavior.